File: min_subarray.py
from typing import List


class Solution:
    def minSubArrayLen(self, target: int, nums: List[int]) -> int:
        left = 0
        smallest = len(nums) + 1
        current = 0

        for right in range(0, len(nums)):
            # if smallest > 0 and right - left >= smallest:
            #     #move to the left
            #     current -= nums[left]
                
            current += nums[right]
            while(current >= target and left <= right):
                smallest = min(smallest, right - left + 1)
                if smallest == 1:
                    return 1
                current -= nums[left]
                left += 1
        return 0 if smallest > len(nums) else smallest
    


if __name__ == '__main__':
    s = Solution()
    print(s.minSubArrayLen(7, [2,3,1,2,4,3]))

File: MinSubarray.java
class Solution {
    public int minSubArrayLen(int target, int[] nums) {
        var left = 0;
        var len = nums.length;
        var smallest = len + 1;
        var current = 0;

        for (int right = 0; right < len; right++) {
            current += nums[right];
            while(current >= target && left <= right) {
                smallest = Math.min(smallest, right - left + 1);
                if (smallest == 1){
                    return 1;
                }
                current -= nums[left++];
                //left += 1
            }
        }
        return smallest > len ? 0 : smallest;
    }

    public static void main(String[] args) {
        System.out.println(new Solution().minSubArrayLen(7, new int[] {2,3,1,2,4,3}));
    }
}